File: frontend/lib/services/api_client.dart
import 'package:dio/dio.dart';

import '../models/crate.dart';
import '../models/purchase_link.dart';
import '../models/refinement.dart';
import '../models/setlist.dart';
import '../models/track_list_response.dart';
import '../providers/setlist_library_provider.dart';

class ApiClient {
  static const _baseUrl = '/api';

  final Dio _dio;

  ApiClient({Dio? dio})
      : _dio = dio ??
            Dio(BaseOptions(
              baseUrl: _baseUrl,
              connectTimeout: const Duration(seconds: 10),
              receiveTimeout: const Duration(seconds: 120),
            ));

  Dio get dio => _dio;

  // -------------------------------------------------------------------------
  // Spotify OAuth
  // -------------------------------------------------------------------------

  Future<bool> checkSpotifyConnection(String userId) async {
    final response = await _dio.get(
      '/auth/spotify/status',
      options: Options(headers: {'X-User-Id': userId}),
    );
    return response.data['connected'] as bool? ?? false;
  }

  Future<String> getSpotifyAuthUrl(String userId) async {
    final response = await _dio.get(
      '/auth/spotify',
      options: Options(headers: {'X-User-Id': userId}),
    );
    return response.data['redirect_url'] as String;
  }

  // -------------------------------------------------------------------------
  // Spotify Import
  // -------------------------------------------------------------------------

  Future<Map<String, dynamic>> importSpotifyPlaylist(
      String playlistUrl) async {
    final response = await _dio.post(
      '/import/spotify',
      data: {'playlist_url': playlistUrl},
    );
    return response.data as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> getImportStatus(String importId) async {
    final response = await _dio.get('/import/$importId');
    return response.data as Map<String, dynamic>;
  }

  // -------------------------------------------------------------------------
  // Track Catalog
  // -------------------------------------------------------------------------

  Future<TrackListResponse> listTracks({
    int page = 1,
    int perPage = 25,
    String sort = 'date_added',
    String order = 'desc',
  }) async {
    final response = await _dio.get('/tracks', queryParameters: {
      'page': page,
      'per_page': perPage,
      'sort': sort,
      'order': order,
    });
    return TrackListResponse.fromJson(response.data as Map<String, dynamic>);
  }

  // -------------------------------------------------------------------------
  // Setlist Generation
  // -------------------------------------------------------------------------

  Future<Setlist> generateSetlist(
    String prompt, {
    int? trackCount,
    String? energyProfile,
    String? sourcePlaylistId,
    String? seedTracklist,
    bool? creativeMode,
    double? bpmMin,
    double? bpmMax,
    bool? verify,
  }) async {
    final data = <String, dynamic>{'prompt': prompt};
    if (trackCount != null) data['track_count'] = trackCount;
    if (energyProfile != null) data['energy_profile'] = energyProfile;
    if (sourcePlaylistId != null) {
      data['source_playlist_id'] = sourcePlaylistId;
    }
    if (seedTracklist != null) data['seed_tracklist'] = seedTracklist;
    if (creativeMode != null) data['creative_mode'] = creativeMode;
    if (bpmMin != null && bpmMax != null) {
      data['bpm_range'] = {'min': bpmMin, 'max': bpmMax};
    }
    if (verify != null) data['verify'] = verify;
    final response = await _dio.post('/setlists/generate', data: data);
    return Setlist.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Setlist> arrangeSetlist(String id, {String? energyProfile}) async {
    final data = <String, dynamic>{};
    if (energyProfile != null) data['energy_profile'] = energyProfile;
    final response = await _dio.post(
      '/setlists/$id/arrange',
      data: data.isNotEmpty ? data : null,
    );
    return Setlist.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Setlist> getSetlist(String id) async {
    final response = await _dio.get('/setlists/$id');
    return Setlist.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<SetlistSummary>> listSetlists() async {
    final response = await _dio.get('/setlists');
    final list = (response.data['setlists'] as List<dynamic>?) ?? [];
    return list
        .map((e) => SetlistSummary.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> deleteSetlist(String id) async {
    await _dio.delete('/setlists/$id');
  }

  Future<Setlist> updateSetlist(String id, {String? name}) async {
    final response = await _dio.patch(
      '/setlists/$id',
      data: {'name': name},
    );
    return Setlist.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Setlist> duplicateSetlist(String id) async {
    final response = await _dio.post('/setlists/$id/duplicate');
    return Setlist.fromJson(response.data as Map<String, dynamic>);
  }

  // -------------------------------------------------------------------------
  // Crates
  // -------------------------------------------------------------------------

  Future<List<Crate>> listCrates() async {
    final response = await _dio.get('/crates');
    final list = (response.data['crates'] as List<dynamic>?) ?? [];
    return list
        .map((e) => Crate.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<Crate> createCrate(String name) async {
    final response = await _dio.post('/crates', data: {'name': name});
    return Crate.fromJson(response.data as Map<String, dynamic>);
  }

  Future<CrateDetail> getCrate(String id) async {
    final response = await _dio.get('/crates/$id');
    return CrateDetail.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteCrate(String id) async {
    await _dio.delete('/crates/$id');
  }

  Future<void> addSetlistToCrate(String crateId, String setlistId) async {
    await _dio.post('/crates/$crateId/setlists/$setlistId');
  }

  Future<void> removeCrateTrack(String crateId, String trackId) async {
    await _dio.delete('/crates/$crateId/tracks/$trackId');
  }

  // -------------------------------------------------------------------------
  // Purchase Links
  // -------------------------------------------------------------------------

  Future<List<PurchaseLink>> getPurchaseLinks({
    required String title,
    required String artist,
  }) async {
    final response = await _dio.get('/purchase-links', queryParameters: {
      'title': title,
      'artist': artist,
    });
    final links = (response.data['links'] as List)
        .map((e) => PurchaseLink.fromJson(e as Map<String, dynamic>))
        .toList();
    return links;
  }

  // -------------------------------------------------------------------------
  // Audio Preview (unified: Deezer → iTunes fallback)
  // -------------------------------------------------------------------------

  /// Unified preview search: backend tries Deezer then iTunes.
  /// Returns a PreviewSearchResult with source, preview URL, and metadata.
  Future<PreviewSearchResult> searchPreview(String title, String artist) async {
    try {
      final response = await _dio.get('/audio/search', queryParameters: {
        'title': title,
        'artist': artist,
      });
      final data = response.data as Map<String, dynamic>;
      return PreviewSearchResult(
        source: data['source'] as String?,
        previewUrl: data['preview_url'] as String?,
        externalUrl: data['external_url'] as String?,
        searchQueries: (data['search_queries'] as List?)
                ?.map((e) => e as String)
                .toList() ??
            [],
        uploaderName: data['uploader_name'] as String?,
      );
    } catch (e) {
      return const PreviewSearchResult(
        source: null,
        previewUrl: null,
        externalUrl: null,
        searchQueries: [],
        uploaderName: null,
      );
    }
  }

  // -------------------------------------------------------------------------
  // Setlist Refinement
  // -------------------------------------------------------------------------

  Future<RefinementResponse> refineSetlist(String setlistId, String message) async {
    final response = await _dio.post(
      '/setlists/$setlistId/refine',
      data: {'message': message},
    );
    return RefinementResponse.fromJson(response.data as Map<String, dynamic>);
  }

  Future<RefinementResponse> revertSetlist(String setlistId, int versionNumber) async {
    final response = await _dio.post('/setlists/$setlistId/revert/$versionNumber');
    return RefinementResponse.fromJson(response.data as Map<String, dynamic>);
  }

  Future<HistoryResponse> getSetlistHistory(String setlistId) async {
    final response = await _dio.get('/setlists/$setlistId/history');
    return HistoryResponse.fromJson(response.data as Map<String, dynamic>);
  }
}

class PreviewSearchResult {
  final String? source;
  final String? previewUrl;
  final String? externalUrl;
  final List<String> searchQueries;
  final String? uploaderName;

  const PreviewSearchResult({
    required this.source,
    required this.previewUrl,
    required this.externalUrl,
    required this.searchQueries,
    this.uploaderName,
  });
}

File: frontend-next/src/lib/api-client.ts
import type {
  Setlist,
  SetlistSummary,
  TrackListResponse,
  Crate,
  CrateDetail,
  PurchaseLink,
  PreviewSearchResult,
  RefinementResponse,
  HistoryResponse,
  ImportStatusResponse,
  ApiError,
} from '@/types';

const BASE_URL = '/api';
const DEFAULT_TIMEOUT = 10_000;
const GENERATE_TIMEOUT = 120_000;

class ApiClientError extends Error {
  code: string;
  constructor(code: string, message: string) {
    super(message);
    this.name = 'ApiClientError';
    this.code = code;
  }
}

async function request<T>(
  path: string,
  options: RequestInit & { timeout?: number; params?: Record<string, string | number> } = {},
): Promise<T> {
  const { timeout = DEFAULT_TIMEOUT, params, ...fetchOptions } = options;

  let url = `${BASE_URL}${path}`;
  if (params) {
    const searchParams = new URLSearchParams();
    for (const [key, value] of Object.entries(params)) {
      searchParams.set(key, String(value));
    }
    url += `?${searchParams.toString()}`;
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout);

  try {
    const response = await fetch(url, {
      ...fetchOptions,
      signal: controller.signal,
      headers: {
        ...(fetchOptions.body ? { 'Content-Type': 'application/json' } : {}),
        ...fetchOptions.headers,
      },
    });

    if (!response.ok) {
      let errorData: ApiError | null = null;
      try {
        errorData = await response.json();
      } catch {
        // non-JSON error
      }
      throw new ApiClientError(
        errorData?.error?.code ?? 'UNKNOWN',
        errorData?.error?.message ?? `Request failed: ${response.status}`,
      );
    }

    // 204 No Content
    if (response.status === 204) {
      return undefined as T;
    }

    return response.json();
  } finally {
    clearTimeout(timer);
  }
}

// ---------------------------------------------------------------------------
// Spotify OAuth
// ---------------------------------------------------------------------------

export async function checkSpotifyConnection(userId: string): Promise<boolean> {
  const data = await request<{ connected: boolean }>('/auth/spotify/status', {
    headers: { 'X-User-Id': userId },
  });
  return data.connected;
}

export async function getSpotifyAuthUrl(userId: string): Promise<string> {
  const data = await request<{ redirect_url: string }>('/auth/spotify', {
    headers: { 'X-User-Id': userId },
  });
  return data.redirect_url;
}

// ---------------------------------------------------------------------------
// Spotify Import
// ---------------------------------------------------------------------------

export async function importSpotifyPlaylist(
  playlistUrl: string,
): Promise<{ import_id: string; status: string }> {
  return request('/import/spotify', {
    method: 'POST',
    body: JSON.stringify({ playlist_url: playlistUrl }),
  });
}

export async function getImportStatus(
  importId: string,
): Promise<ImportStatusResponse> {
  return request(`/import/${importId}`);
}

// ---------------------------------------------------------------------------
// Track Catalog
// ---------------------------------------------------------------------------

export async function listTracks(options?: {
  page?: number;
  perPage?: number;
  sort?: string;
  order?: string;
}): Promise<TrackListResponse> {
  return request('/tracks', {
    params: {
      page: options?.page ?? 1,
      per_page: options?.perPage ?? 25,
      sort: options?.sort ?? 'date_added',
      order: options?.order ?? 'desc',
    },
  });
}

// ---------------------------------------------------------------------------
// Setlist Generation
// ---------------------------------------------------------------------------

export async function generateSetlist(options: {
  prompt: string;
  trackCount?: number;
  energyProfile?: string;
  sourcePlaylistId?: string;
  seedTracklist?: string;
  creativeMode?: boolean;
  bpmMin?: number;
  bpmMax?: number;
  verify?: boolean;
}): Promise<Setlist> {
  const body: Record<string, unknown> = { prompt: options.prompt };
  if (options.trackCount != null) body.track_count = options.trackCount;
  if (options.energyProfile != null) body.energy_profile = options.energyProfile;
  if (options.sourcePlaylistId != null) body.source_playlist_id = options.sourcePlaylistId;
  if (options.seedTracklist != null) body.seed_tracklist = options.seedTracklist;
  if (options.creativeMode != null) body.creative_mode = options.creativeMode;
  if (options.bpmMin != null && options.bpmMax != null) {
    body.bpm_range = { min: options.bpmMin, max: options.bpmMax };
  }
  if (options.verify != null) body.verify = options.verify;

  return request('/setlists/generate', {
    method: 'POST',
    body: JSON.stringify(body),
    timeout: GENERATE_TIMEOUT,
  });
}

export async function arrangeSetlist(
  id: string,
  energyProfile?: string,
): Promise<Setlist> {
  const body: Record<string, unknown> = {};
  if (energyProfile) body.energy_profile = energyProfile;

  return request(`/setlists/${id}/arrange`, {
    method: 'POST',
    body: Object.keys(body).length > 0 ? JSON.stringify(body) : undefined,
  });
}

export async function getSetlist(id: string): Promise<Setlist> {
  return request(`/setlists/${id}`);
}

export async function listSetlists(): Promise<SetlistSummary[]> {
  const data = await request<{ setlists: SetlistSummary[] }>('/setlists');
  return data.setlists ?? [];
}

export async function deleteSetlist(id: string): Promise<void> {
  return request(`/setlists/${id}`, { method: 'DELETE' });
}

export async function updateSetlist(
  id: string,
  name: string,
): Promise<Setlist> {
  return request(`/setlists/${id}`, {
    method: 'PATCH',
    body: JSON.stringify({ name }),
  });
}

export async function duplicateSetlist(id: string): Promise<Setlist> {
  return request(`/setlists/${id}/duplicate`, { method: 'POST' });
}

// ---------------------------------------------------------------------------
// Crates
// ---------------------------------------------------------------------------

export async function listCrates(): Promise<Crate[]> {
  const data = await request<{ crates: Crate[] }>('/crates');
  return data.crates ?? [];
}

export async function createCrate(name: string): Promise<Crate> {
  return request('/crates', {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
}

export async function getCrate(id: string): Promise<CrateDetail> {
  return request(`/crates/${id}`);
}

export async function deleteCrate(id: string): Promise<void> {
  return request(`/crates/${id}`, { method: 'DELETE' });
}

export async function addSetlistToCrate(
  crateId: string,
  setlistId: string,
): Promise<void> {
  return request(`/crates/${crateId}/setlists/${setlistId}`, {
    method: 'POST',
  });
}

export async function removeCrateTrack(
  crateId: string,
  trackId: string,
): Promise<void> {
  return request(`/crates/${crateId}/tracks/${trackId}`, {
    method: 'DELETE',
  });
}

// ---------------------------------------------------------------------------
// Purchase Links
// ---------------------------------------------------------------------------

export async function getPurchaseLinks(
  title: string,
  artist: string,
): Promise<PurchaseLink[]> {
  const data = await request<{ links: PurchaseLink[] }>('/purchase-links', {
    params: { title, artist },
  });
  return data.links;
}

// ---------------------------------------------------------------------------
// Audio Preview
// ---------------------------------------------------------------------------

export async function searchPreview(
  title: string,
  artist: string,
): Promise<PreviewSearchResult> {
  try {
    return await request<PreviewSearchResult>('/audio/search', {
      params: { title, artist },
    });
  } catch {
    return {
      source: null,
      preview_url: null,
      external_url: null,
      search_queries: [],
      uploader_name: null,
      spotify_uri: null,
    };
  }
}

// ---------------------------------------------------------------------------
// Refinement
// ---------------------------------------------------------------------------

export async function refineSetlist(
  setlistId: string,
  message: string,
): Promise<RefinementResponse> {
  return request(`/setlists/${setlistId}/refine`, {
    method: 'POST',
    body: JSON.stringify({ message }),
    timeout: GENERATE_TIMEOUT,
  });
}

export async function revertSetlist(
  setlistId: string,
  versionNumber: number,
): Promise<RefinementResponse> {
  return request(`/setlists/${setlistId}/revert/${versionNumber}`, {
    method: 'POST',
  });
}

export async function getSetlistHistory(
  setlistId: string,
): Promise<HistoryResponse> {
  return request(`/setlists/${setlistId}/history`);
}

export { ApiClientError };
